File: pydsopp/storage/track_storage.py
import sys, os

sys.path.insert(0, os.path.abspath(os.path.join("${STORAGE_PATH}", "proto")))

import agent_settings_pb2 as agent_settings
import ecef_poses_pb2 as ecef_poses
import frame_pb2 as frame
import gnss_track_pb2 as gnss_track
import connection_pb2 as connection
import sanity_check_results_pb2 as sanity_check_results

LONG_UNSIGNED_SIZE = 8
UNSIGNED_SIZE = 4


## Saves one proto message
#  @param file opened file for writing
#  @param proto_message protobuf container
def saveProto(file, proto_message):

    message_size = proto_message.ByteSize()
    message_size_bytes = message_size.to_bytes(UNSIGNED_SIZE,
                                               byteorder='little')
    file.write(message_size_bytes)
    file.write(proto_message.SerializeToString())


## Reads one proto message
#  @param file opened file for reading
#  @param proto_message protobuf container
#  @return proto message
def readProto(file, proto_message):

    message_size_bytes = file.read(UNSIGNED_SIZE)
    message_size = int.from_bytes(message_size_bytes, byteorder='little')
    proto_message_bytes = file.read(message_size)
    proto_message.ParseFromString(proto_message_bytes)
    return proto_message


## Class for storing data
class TrackStorage:

    ## Saves data in file
    #  @param self object pointer
    #  @param filename filename string
    #  @return
    def save(self, filename):

        file = open(filename, 'wb')
        number_of_frames = len(self.frames)
        number_of_frames_bytes = number_of_frames.to_bytes(LONG_UNSIGNED_SIZE,
                                                           byteorder='little')
        file.write(number_of_frames_bytes)
        for frame in self.frames:
            saveProto(file, frame)

        saveProto(file, self.connections)
        saveProto(file, self.gnss_track)
        saveProto(file, self.ecef_poses)
        saveProto(file, self.sanity_check_results)
        saveProto(file, self.agent_settings)

        file.close()

    ## Reads data from file
    #  @param self object pointer
    #  @param filename filename string
    #  @return
    def read(self, filename):

        file = open(filename, 'rb')

        number_of_frames_bytes = file.read(LONG_UNSIGNED_SIZE)
        number_of_frames = int.from_bytes(number_of_frames_bytes,
                                          byteorder='little')

        self.frames = [
            readProto(file, frame.Keyframe()) for i in range(number_of_frames)
        ]
        self.connections = readProto(file, connection.Connections())
        self.gnss_track = readProto(file, gnss_track.GnssTrack())
        self.ecef_poses = readProto(file, ecef_poses.ECEFPoses())
        self.sanity_check_results = readProto(
            file, sanity_check_results.SanityCheckResults())
        self.agent_settings = readProto(file, agent_settings.AgentSettings())

        file.close()

    ## @var frames
    #  protobuf containers for frames
    ## @var connections
    #  protobuf container for connections
    ## @var gnss_track
    #  protobuf container for gnss track
    ## @var ecef_poses
    #  protobuf container for ecef poses
    ## @var sanity_check_results
    #  protobuf container for sanity check results
    ## @var agent_settings
    #  protobuf container for agent settings

File: src/storage/src/track_storage.cpp
#include "storage/track_storage.hpp"

#include <fstream>
#include <vector>

#include "agent/agent_settings.hpp"
#include "common/settings.hpp"

namespace dsopp::track::storage {

namespace {

template <class ProtoMessage>
void saveProto(std::ostream &stream, const ProtoMessage &proto_message) {
#if GOOGLE_PROTOBUF_VERSION >= 3010000
  unsigned message_size = static_cast<unsigned>(proto_message.ByteSizeLong());
#else
  unsigned message_size = static_cast<unsigned>(proto_message.ByteSize());
#endif
  stream.write(reinterpret_cast<char *>(&message_size), sizeof(unsigned));
  proto_message.SerializeToOstream(&stream);
}

template <class ProtoMessage>
bool readProto(std::istream &stream, ProtoMessage &proto_message) {
  unsigned message_size;
  stream.read(reinterpret_cast<char *>(&message_size), sizeof(unsigned));
  std::vector<char> message_buffer(message_size);
  stream.read(message_buffer.data(), message_size);
  if (!proto_message.ParseFromArray(message_buffer.data(), static_cast<int>(message_size))) return false;
  return true;
}

}  // namespace

storage::Track::Track() = default;

Track::Track(const proto::GnssTrack &gnss_track, const proto::ECEFPoses &ecef_poses,
             const std::map<size_t, sanity_checker::SanityCheckStatus> &sanity_check_results,
             const dsopp::sensors::proto::AgentSettings &agent_settings, const OdometryTrack &odometry_track) {
  this->gnss_track_ = gnss_track;
  this->ecef_poses_ = ecef_poses;
  for (const auto &[frame_index, result] : sanity_check_results) {
    (*this->sanity_check_results_.mutable_sanity_check_results())[frame_index] = proto::SanityCheckStatus(result);
  }
  this->agent_settings_ = agent_settings;
  this->odometry_track_ = odometry_track;
}

void storage::Track::save(std::ostream &stream) const {
  long unsigned number_of_frames = this->odometry_track_.frames.size();
  stream.write(reinterpret_cast<char *>(&number_of_frames), sizeof(long unsigned));
  for (auto &frame : this->odometry_track_.frames) saveProto(stream, frame);

  saveProto(stream, this->odometry_track_.connections);
  saveProto(stream, this->gnss_track_);
  saveProto(stream, this->ecef_poses_);
  saveProto(stream, this->sanity_check_results_);
  saveProto(stream, this->agent_settings_);
}

bool storage::Track::read(std::istream &stream) {
  long unsigned number_of_frames;
  stream.read(reinterpret_cast<char *>(&number_of_frames), sizeof(long unsigned));
  this->odometry_track_.frames.resize(number_of_frames);
  for (size_t i = 0; i < number_of_frames; ++i) {
    if (!readProto(stream, this->odometry_track_.frames[i])) return false;
  }

  if (!readProto(stream, this->odometry_track_.connections)) return false;
  if (!readProto(stream, this->gnss_track_)) return false;
  if (!readProto(stream, this->ecef_poses_)) return false;
  if (!readProto(stream, this->sanity_check_results_)) return false;
  if (!readProto(stream, this->agent_settings_)) return false;
  return true;
}

const proto::GnssTrack &storage::Track::gnssTrack() const { return this->gnss_track_; }

const proto::ECEFPoses &storage::Track::ecefPoses() const { return this->ecef_poses_; }

const proto::SanityCheckResults &storage::Track::sanityCheckResults() const { return this->sanity_check_results_; }

const dsopp::sensors::proto::AgentSettings &storage::Track::agentSettings() const { return this->agent_settings_; }

const OdometryTrack &storage::Track::odometryTrack() const { return this->odometry_track_; }
}  // namespace dsopp::track::storage
